crates/zizmor/src/audit/ref_version_mismatch.rs

```diff
@@ -0,0 +1,130 @@
+use std::sync::LazyLock;
+
+use anyhow::{Result, anyhow};
+use github_actions_models::common::Uses;
+use regex::Regex;
+
+use crate::{
+    audit::{Audit, AuditLoadError, AuditState, audit_meta},
+    finding::{
+        Confidence, Finding, Severity,
+        location::{Comment, Locatable},
+    },
+    github_api,
+    models::{StepCommon, uses::RepositoryUsesExt, workflow::Step},
+};
+
+pub(crate) struct RefVersionMismatch {
+    client: github_api::Client,
+}
+
+audit_meta!(
+    RefVersionMismatch,
+    "ref-version-mismatch",
+    "detects commit SHAs that don't match their version comment tags"
+);
+
+static VERSION_COMMENT_PATTERNS: LazyLock<Vec<Regex>> = LazyLock::new(|| {
+    vec![
+        // Matches "# tag=v2.8.0" or "# tag=v1.2.3"
+        Regex::new(r"#\s*tag\s*=\s*(v\d+(?:\.\d+)*(?:\.\d+)?)").unwrap(),
+        // Matches "# v2.8.0"
+        Regex::new(r"#\s*(v\d+(?:\.\d+)*(?:\.\d+)?)").unwrap(),
+        // Matches version without 'v' prefix: "# tag=2.8.0"
+        Regex::new(r"#\s*tag\s*=\s*(\d+(?:\.\d+)*(?:\.\d+)?)").unwrap(),
+        // More flexible: "# version: 2.8.0"
+        Regex::new(r"#\s*(?:version|ver)\s*[:=]\s*(v?\d+(?:\.\d+)*(?:\.\d+)?)").unwrap(),
+    ]
+});
+
+impl RefVersionMismatch {
+    fn extract_version_from_comments<'doc>(
+        &self,
+        comments: &'doc Vec<Comment<'doc>>,
+    ) -> Option<&'doc str> {
+        for comment in comments {
+            for pattern in VERSION_COMMENT_PATTERNS.iter() {
+                if let Some(captures) = pattern.captures(comment.as_ref()) {
+                    if let Some(version_match) = captures.get(1) {
+                        return Some(version_match.as_str());
+                    }
+                }
+            }
+        }
+        None
+    }
+}
+
+impl Audit for RefVersionMismatch {
+    fn new(state: &AuditState<'_>) -> Result<Self, AuditLoadError> {
+        if state.no_online_audits {
+            return Err(AuditLoadError::Skip(anyhow!(
+                "offline audits only requested"
+            )));
+        }
+
+        let Some(client) = state.github_client() else {
+            return Err(AuditLoadError::Skip(anyhow!(
+                "can't run without a GitHub API token"
+            )));
+        };
+
+        Ok(Self { client })
+    }
+
+    fn audit_step<'doc>(&self, step: &Step<'doc>) -> anyhow::Result<Vec<Finding<'doc>>> {
+        let mut findings = vec![];
+
+        let Some(Uses::Repository(uses)) = step.uses() else {
+            return Ok(findings);
+        };
+
+        // Only check steps that have commit refs (not symbolic refs like v1.0.0)
+        let Some(commit_sha) = uses.commit_ref() else {
+            return Ok(findings);
+        };
+
+        // For now, we'll search the entire document source for version patterns
+        // that might correspond to this uses line. This is a simplified approach.
```

Comment:
Q: is this comment accurate still? It looks like the implementation below doesn't scan the whole doc source, only the comment next to the `uses:` clause (which is the right way to do it!)